dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/config/DefaultGatewayAdministrationService.java

```diff
@@ -121,58 +115,62 @@ public boolean addGateway(SmsGatewayConfig config) {
   }
 
   @Override
-  public void updateGateway(SmsGatewayConfig persistedConfig, SmsGatewayConfig updatedConfig) {
-    initState();
-
-    if (persistedConfig == null || updatedConfig == null) {
-      log.warn("Gateway configurations cannot be null");
-      return;
+  public void updateGateway(
+      @CheckForNull SmsGatewayConfig persisted, @CheckForNull SmsGatewayConfig updated)
+      throws NotFoundException, ConflictException {
+    if (updated == null) throw new ConflictException("Gateway configuration cannot be null");
+    if (persisted == null) throw new NotFoundException(SmsGatewayConfig.class, updated.getUid());
+    if (persisted.getClass() != updated.getClass())
+      throw new ConflictException("Type of an existing configuration cannot be changed");
+
+    updated.setUid(persisted.getUid());
+    updated.setDefault(persisted.isDefault());
+
+    if (updated.getPassword() == null) {
+      // keep old password when undefined
+      updated.setPassword(persisted.getPassword());
+    } else if (persisted.getPassword() != null
+        && !persisted.getPassword().equals(updated.getPassword())) {
+      // password change
+      updated.setPassword(pbeStringEncryptor.encrypt(updated.getPassword()));
     }
 
-    updatedConfig.setUid(persistedConfig.getUid());
-    updatedConfig.setDefault(persistedConfig.isDefault());
-
-    if (persistedConfig.getPassword() != null
-        && !persistedConfig.getPassword().equals(updatedConfig.getPassword())) {
-      updatedConfig.setPassword(pbeStringEncryptor.encrypt(updatedConfig.getPassword()));
+    if (persisted instanceof ClickatellGatewayConfig
+        && updated instanceof ClickatellGatewayConfig) {
+      ClickatellGatewayConfig from = (ClickatellGatewayConfig) persisted;
+      ClickatellGatewayConfig to = (ClickatellGatewayConfig) updated;
+      if (to.getAuthToken() == null) to.setAuthToken(from.getAuthToken());
     }
 
-    if (persistedConfig instanceof GenericHttpGatewayConfig) {
-      List<GenericGatewayParameter> newList = new ArrayList<>();
-
-      GenericHttpGatewayConfig persistedGenericConfig = (GenericHttpGatewayConfig) persistedConfig;
-      GenericHttpGatewayConfig updatedGenericConfig = (GenericHttpGatewayConfig) updatedConfig;
-
-      List<GenericGatewayParameter> persistedList =
-          persistedGenericConfig.getParameters().stream()
-              .filter(GenericGatewayParameter::isConfidential)
-              .collect(Collectors.toList());
-
-      List<GenericGatewayParameter> updatedList =
-          updatedGenericConfig.getParameters().stream()
-              .filter(GenericGatewayParameter::isConfidential)
-              .collect(Collectors.toList());
-
-      for (GenericGatewayParameter p : updatedList) {
-        if (!isPresent(persistedList, p)) {
-          p.setValue(pbeStringEncryptor.encrypt(p.getValue()));
+    if (persisted instanceof GenericHttpGatewayConfig
+        && updated instanceof GenericHttpGatewayConfig) {
```

Comment:
just a small suggestion, this instanceof check always gets executed even if the instanceof check on line 138 is true. Maybe some reorganisation of the code could allow not checking when we already know the answer.